File: typical_content001/dfs.rb
if !ENV['RUBY_THREAD_VM_STACK_SIZE']
  require 'rubygems';RUBY=Gem.ruby
  exec({'RUBY_THREAD_VM_STACK_SIZE'=>'100000000'},RUBY,$0)
end
$h,$w = gets.split.map(&:to_i)
$visited = {}
$cs = []
$h.times do
  $cs << gets.chomp.chars
end

def dfs(i, j)
  return if i < 0 || j < 0
  return if i >= $h || j >= $w
  return if $cs[i][j] == '#'
  $visited[i] ||= {}
  return if $visited[i][j]
  $visited[i][j] = true
  if $cs[i][j] == 'g'
    puts 'Yes'
    exit
  end
  dfs(i + 1, j)
  dfs(i - 1, j)
  dfs(i, j + 1)
  dfs(i, j - 1)
end

$cs.each.with_index do |row,i|
  row.each.with_index do |target,j|
    if target == 's'
      dfs(i,j)
    end
  end
end

puts 'No'

File: typical_content001/dfs.cpp
#include <iostream>
using namespace std;

int H, W, visited[500][500];
char c[500][500];

void dfs(int x, int y){
  if(x < 0 || y < 0){
    return;
  }
  if(x >= W || y >= H){
    return;
  }
  if(visited[x][y] == 1){
    return;
  }
  if(c[x][y] == '#'){
    return;
  }
  visited[x][y] = 1;
  dfs(x+1,y);
  dfs(x-1,y);
  dfs(x,y+1);
  dfs(x,y-1);
};

int main(){
  int i,j,xs,ys,xg,yg;
	cin >> H >> W;

  for(i=0; i<W; i++){
    for(j=0; j<H; j++){
      cin >> c[i][j];
      if(c[i][j] == 's'){
        xs = i;
        ys = j;
      }
      if(c[i][j] == 'g'){
        xg = i;
        yg = j;
      }
    }
  }

  dfs(xs,ys);

  if(visited[xg][yg] == 1){
    cout << "Yes" << endl;
  }else{
    cout << "No" << endl;
  }
}
